### PR TITLE
Pipeline item skip mechanism.

### LINKED ISSUE #53: Design item skip pattern in Pipeline
Pipeline should be able to skip items. This can be achieved in mulitple ways:

- Throwing exception.
- Calling `stopPropagation()` and detecting it.

Should we throw a `.skip` event in that case?

### PR DESCRIPTION
**Pull request by GrandLTU**
Closes #53.

### CODE REVIEW

**Comment by lmikelionis** [comment]
its no an issue with your code, but maybe you can update LICESNSE file with same PR?

File: EventListener/AbstractConsumeEventListener.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ONGR\ConnectionsBundle\EventListener;
+
+use ONGR\ConnectionsBundle\Pipeline\Event\ItemPipelineEvent;
+
+/**
+ * AbstractConsumeEventListener class.
```
Comment (tautrimas) [inline EventListener/AbstractConsumeEventListener.php]: Should describe responsibility of the class.

File: Pipeline/ItemSkipException.php [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+ * (c) NFQ Technologies UAB <info@nfq.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace ONGR\ConnectionsBundle\Pipeline;
+
+/**
+ * Skip Item Exception.
```
Comment (tautrimas) [inline Pipeline/ItemSkipException.php]: I suggest describing, when this exception should be used.

File: Tests/Unit/Pipeline/PipelineTest.php [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+namespace ONGR\ConnectionsBundle\Tests\Unit\Pipeline;
+
+use ONGR\ConnectionsBundle\Pipeline\Event\ItemPipelineEvent;
+use ONGR\ConnectionsBundle\Pipeline\Event\SourcePipelineEvent;
+use ONGR\ConnectionsBundle\Pipeline\ItemSkipException;
+use ONGR\ConnectionsBundle\Pipeline\PipelineFactory;
+use Symfony\Component\EventDispatcher\EventDispatcher;
+
+/**
+ * PipelineTest class.
```
Comment (tautrimas) [inline Tests/Unit/Pipeline/PipelineTest.php]: Unnecessary comment.

File: Tests/Unit/Pipeline/PipelineTest.php [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            throw new ItemSkipException();
+        }
+    }
+
+    /**
+     * Pipeline data provider.
+     *
+     * @return array
+     */
+    public function pipelineData()
```
Comment (tautrimas) [inline Tests/Unit/Pipeline/PipelineTest.php]: Data providers must go above their tests (see: http://ongr.readthedocs.org/en/latest/handbook/contributing/coding_standards.html#structure)

File: Tests/Unit/Pipeline/PipelineTest.php [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    {
+        return [
+            [[], 0, 0],
```
Comment (tautrimas) [inline Tests/Unit/Pipeline/PipelineTest.php]: I suggest writing initial and expected condition comments in format `// Case #0: when something, then should skip 0 items.`

File: Tests/Unit/Pipeline/PipelineTestConsumer.php [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+     * @var int
+     */
+    private $skipCalled = 0;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function consume(ItemPipelineEvent $event)
+    {
+        ++$this->consumeCalled;
```
Comment (tautrimas) [inline Tests/Unit/Pipeline/PipelineTestConsumer.php]: Maybe use more standard suffix `$a++` variant?

File: Tests/Unit/Pipeline/PipelineTestConsumer.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function skip(ItemPipelineEvent $event)
+    {
+        ++$this->skipCalled;
```
Comment (tautrimas) [inline Tests/Unit/Pipeline/PipelineTestConsumer.php]: Maybe use more standard suffix `$a++` variant?

File: Pipeline/Event/ItemPipelineEvent.php [hunk 8]
```diff
@@ -31,6 +32,11 @@ class ItemPipelineEvent extends Event
     private $output;
 
     /**
+     * @var ItemSkipException
+     */
+    private $itemSkipException;
```
Comment (tautrimas) [inline Pipeline/Event/ItemPipelineEvent.php]: Maybe simply `$skipException`?

**Pull request merged**